src/apply.ts

```diff
@@ -26,154 +24,85 @@ interface ReadOptions {
+	const bundle = await resourceBundle.read<SDK.Release>(
 		options.stream,
 		'io.balena.release',
 	);
 
+	// TODO: validate if manifest version is compatible with SDK
+
+	let release: Release = { releaseImages: [] };
+	try {
+		release = normalizeManifest(bundle.manifest);
+	} catch (error) {
+		throw new Error(`Manifest is malformed: ${error.message}`);
+	}
+
 	const sdk = SDK.getSdk({
 		apiUrl: options.apiUrl,
 		dataDirectory: false,
 	});
 	await sdk.auth.loginWithToken(options.authToken);
 
-	// TODO: validate manifest
-	// Check if a release with the same version and revision exists
-	const [existingRelease] = await sdk.pine.get({
+	const existingReleases = await sdk.pine.get({
 		resource: 'release',
 		options: {
-			$select: [
-				'id',
-				'is_final',
-				'is_invalidated',
-				'revision',
-				'semver',
-				'status',
-				'version',
-			],
+			$select: ['id'],
 			$filter: {
 				belongs_to__application: options.application,
-				semver: bundle.manifest.semver,
-				revision: bundle.manifest.revision,
+				semver_major: bundle.manifest.semver_major,
+				semver_minor: bundle.manifest.semver_minor,
+				semver_patch: bundle.manifest.semver_patch,
+				status: 'success',
 			},
+			$top: 1,
 		},
 	});
 
-	if (existingRelease) {
-		if (!options.force) {
-			throw new Error(
-				'An application release with the same version already exists.',
-			);
-		}
-		// Remove images mapped to a release
-		// when the following conditions about the release is true
-		// - the release is is_final is false and/or is_invalidated is true
-		// - force flag is set to true
-		type ReleaseImageMap = {
-			id: number;
-			created_at: string;
-			image: {
-				__id: number;
-			};
-			is_part_of__release: {
-				__id: number;
-			};
-		};
-		const releaseImageMappings = (await sdk.pine.get({
-			resource: 'image__is_part_of__release',
-			options: {
-				$filter: {
-					is_part_of__release: existingRelease.id,
-				},
-			},
-		})) as unknown as ReleaseImageMap[];
-
-		await Promise.all(
-			releaseImageMappings.map(async (releaseImageMap) => {
-				await sdk.pine.delete({
-					resource: 'image__is_part_of__release',
-					id: releaseImageMap.id,
-				});
-			}),
-		);
-
-		// TODO: Delete images tied to the old release
-		// TODO: Remove release tags of old release
-	}
-
-	// Backfill lower releases to workaround the balena API constraint
-	// that a particular release revision may only be created if
-	// lower revisions exist
-	// TODO: Remove backfilling once balena API constraint is removed
-	if (bundle.manifest.revision != null) {
-		await backfillLowerRevisions(
-			sdk,
-			options.application,
-			currentDateIso,
-			bundle.manifest,
+	if (existingReleases.length > 0) {
+		// State the what and explain the why.  We need to say what constraint the user is hitting.
+		throw new Error(
+			'A successful release with this version already exists; only 1 successfuly release with the same version is allowed.',
```

Comment:
I think a slightly better error message is:

```suggestion
			'A release with version ${release.semver} already exists and duplicates are not allowed.',
```

Reply:
The semver includes the revision number which is irrelevant to us.  We can combine the major, minor, and patch versions instead:
```
		const version = `${bundle.manifest.semver_major}.${bundle.manifest.semver_minor}.${bundle.manifest.semver_patch}`;
		throw new Error(
			`A successful release with the version ${version} already exists and duplicates are not allowed.`,
		);
```